# ==============================================================================
# REGRESSOR COMMIT DIFF
# ==============================================================================
# File: dom/serviceworkers/ServiceWorkerScriptCache.cpp
# Commit: 0e68db4ad6af
# Full Hash: 0e68db4ad6afe1f43d97533316d80e5aadd96082
# Author: Luca Greco <lgreco@mozilla.com>
# Date: 2020-07-01 21:32:12
# Regressor Bug: 1609920
# File Overlap Count: 2
# Description:
#   Bug 1609920 - part 1: Allow the WebExtension Framework to register a moz-extension service worker. r=dom-workers-and-storage-reviewers,asuth,mixedpuppy
#   
#   - Adds the new about:config pref "extensions.backgroundServiceWorker.enabled" (currently defaults to false).
#   - Adds the background.service_worker property to the manifest JSON schema definition
#   - Locks background.service_worker manifest property behind the new preference
# ==============================================================================

diff -r 2a107eabe00c -r 0e68db4ad6af dom/serviceworkers/ServiceWorkerScriptCache.cpp
--- a/dom/serviceworkers/ServiceWorkerScriptCache.cpp	Mon Jun 29 16:18:44 2020 +0000
+++ b/dom/serviceworkers/ServiceWorkerScriptCache.cpp	Wed Jun 10 12:13:29 2020 +0000
@@ -882,6 +882,54 @@
     return NS_OK;
   }
 
+  nsCOMPtr<nsIChannel> channel = do_QueryInterface(request);
+  MOZ_ASSERT(channel, "How come we don't have any channel?");
+
+  // TODO: also check a pref and disallow if the pref is false.
+  nsCOMPtr<nsIURI> uri;
+  channel->GetOriginalURI(getter_AddRefs(uri));
+  bool isExtension = uri->SchemeIs("moz-extension");
+
+  if (isExtension) {
+    // TODO: avoid duplicated parts that could be shared with the HTTP channel
+    // scenario?
+    // TODO: check aStatus and early exit if loading the script body failed?
+    nsCOMPtr<nsIURI> channelURL;
+    rv = channel->GetURI(getter_AddRefs(channelURL));
+    if (NS_WARN_IF(NS_FAILED(rv))) {
+      return rv;
+    }
+
+    nsCString channelURLSpec;
+    MOZ_ALWAYS_SUCCEEDS(channelURL->GetSpec(channelURLSpec));
+
+    MOZ_DIAGNOSTIC_ASSERT(!mURLList.IsEmpty());
+    if (channelURLSpec != mURLList[0]) {
+      // Append the final URL (a file or jar url).
+      mURLList.AppendElement(channelURLSpec);
+    }
+
+    // TODO: refreshing a moz-extension service worker should only
+    // happen through the addon updates?
+    //
+    // NOTE: trying to register any moz-extension use that doesn't ends
+    // with .js/.jsm/.mjs seems to be already completing with an error
+    // in aStatus and they never reach this point.
+    char16_t* buffer = nullptr;
+    size_t len = 0;
+    rv = ScriptLoader::ConvertToUTF16(channel, aString, aLen,
+                                      NS_LITERAL_STRING("UTF-8"), nullptr,
+                                      buffer, len);
+    if (NS_WARN_IF(NS_FAILED(rv))) {
+      return rv;
+    }
+
+    mBuffer.Adopt(buffer, len);
+
+    rv = NS_OK;
+    return NS_OK;
+  }
+
   nsCOMPtr<nsIHttpChannel> httpChannel = do_QueryInterface(request);
   MOZ_ASSERT(httpChannel, "How come we don't have an HTTP channel?");
 